shoov/modules/custom/shoov_ci_build/shoov_ci_build.module

```diff
@@ -58,6 +69,81 @@ function  shoov_ci_build_trigger_ci_build_item($message, $ajax = 'nojs') {
 }
 
 /**
+ * Check permissions to un/subscribe account to the CI Build.
+ *
+ * @param $ci_build
+ *  The CI Build node object
+ * @param $account
+ *  The user object.
+ * @param $token
+ *  The access token
+ *
+ * @return bool
+ *  Returns TRUE if user has permissions to un/subscribe account to the CI Build.
+ */
+function shoov_ci_build_subscribe_user_check_access($ci_build, $account, $token) {
+  if (!user_access('administer flags')) {
+    // User doesn't have permissions to un/subscribe account.
+    return FALSE;
+  }
+
+  if (md5($ci_build->nid . '+' . $account->uid) != $token) {
+    // Access token is invalid.
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+ * Un/subscribe user to the CI Build.
+ *
+ * @param $ci_build
+ *  The CI Build node.
+ * @param $account
+ *  The user object.
+ * @param $action
+ *  Action to do. Can be 'subscribe' or 'unsubscribe'
+ * @param $ajax
+ *  Indicate if ajax should be used. Default to 'nojs'.
+ */
+function shoov_ci_build_subscribe_user($ci_build, $account, $action, $ajax = 'nojs') {
+  // Get the flag.
+  $flag = flag_get_flag('subscribe_ci_builds');
+  $flag_action = $action == 'subscribe' ? 'flag' : 'unflag';
+  // Un/flag user due to the sent action.
+  $flag->flag($flag_action, $ci_build->nid, $account);
+
+  if (!$ajax) {
+    $params = array(
+      '@uid' => $account->uid,
+      '@action' => $action == 'subscribe' ? 'subscribed to' : 'unsubscribed from',
+      '@build_id' => $ci_build->nid,
+    );
+    drupal_set_message(t('User @uid has been @action the CI Build @build_id', $params));
+    drupal_goto('<front>');
+  }
+
+  $commands = array();
+  // Change the link to the opposite.
+  if ($action == 'subscribe') {
+    $commands[] = ajax_command_replace('#subscribe-' . $account->uid,
+      l(t('Unsubscribe'), 'ci_build/' . $ci_build->nid . '/' . $account->uid . '/unsubscribe/nojs/' . md5($ci_build->nid . '+' . $account->uid),
```

Comment:
@amitaibu Is it better to send the token also to the page callback and reuse it? Or recalculate it again?